libfreerdp/core/gateway/rdg.c

```diff
@@ -1335,7 +1336,8 @@ static BOOL rdg_process_handshake_response(rdpRdg* rdg, wStream* s)
 	         ", extendedAuth=%s",
 	         error, verMajor, verMinor, serverVersion, extended_auth_to_string(extendedAuth));
 
-	if (FAILED(errorCode))
+	/* [MS-TSGU] 3.7.5.2.3 */
+	if (errorCode != 0)
```

Comment:
according to `[MS-RSGU]` this is an `UNSIGNED` but a `HRESULT`, go figure...

Reply:
I think the old code was correct - `FAILED` will do the correct casting (see https://learn.microsoft.com/en-us/windows/win32/api/winerror/nf-winerror-failed).

Reply:
```
grep -r '#define FAILED'
winpr/include/winpr/error.h:#define FAILED(hr) (((hr)) < 0)
```
I noted that FAILED never triggered because it did not disconnect when getting a errorCode=E_PROXY_COOKIE_AUTHENTICATION_ACCESS_DENIED

Reply:
@akarl10 Yes - obviously our `FAILED` implementation is not correct

Reply:
> according to `[MS-RSGU]` this is an `UNSIGNED` but a `HRESULT`, go figure...

i just looked what is used in that specific case and then copied the allowed values from [MS-TSGU]
This is why sometimes I have written != 0 and sometime != S_OK

Reply:
Here is a description of the `HRESULT` type: https://learn.microsoft.com/en-us/office/client-developer/outlook/mapi/hresult

Basically it is a 32-bit value that encodes various kinds of return values (errors, success, warnings, ...)

`HRESULT` is defined as a signed type so you can easily check for failure (most-significant bit is set - so the value is negative) or success (value is `>= 0`). Checking for `!= 0` is not 100% correct therefore.

Reply:
> I think the old code was correct - `FAILED` will do the correct casting (see https://learn.microsoft.com/en-us/windows/win32/api/winerror/nf-winerror-failed).

well, I can change the define in winpr and leave only the `errorCode != S_OK && errorCode != E_PROXY_QUARANTINE_ACCESSDENIED` case

Reply:
> > I think the old code was correct - `FAILED` will do the correct casting (see https://learn.microsoft.com/en-us/windows/win32/api/winerror/nf-winerror-failed).
> 
> well, I can change the define in winpr and leave only the `errorCode != S_OK && errorCode != E_PROXY_QUARANTINE_ACCESSDENIED` case

I think that's the proper fix - but the check should remain `if (FAILED(errorCode))` or do you need any special handling for `E_PROXY_QUARANTINE_ACCESSDENIED`?

Reply:
> 

generally speaking, yes, negative values are errors, positive successes.
in this case tough (https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-tsgu/390deecf-4c75-4838-8be5-64be3083d3eb) the exact values that are success are defined

Reply:
@akarl10 no, do not touch the define, that is correct. you need to read the value into a `HRESULT`, so just change the definition of the variable and you should be fine.

> > I think the old code was correct - `FAILED` will do the correct casting (see https://learn.microsoft.com/en-us/windows/win32/api/winerror/nf-winerror-failed).
> 
> well, I can change the define in winpr and leave only the `errorCode != S_OK && errorCode != E_PROXY_QUARANTINE_ACCESSDENIED` case